[oshift] Collect ClusterResourceQuota metrics

xvello opened the pull request:
### What does this PR do?

OpenShift allows cluster admins to set cluster-wide quotas for a set of namespaces/projects. This PR adds support for the [`ClusterResourceQuota`](https://docs.openshift.org/3.9/rest_api/apis-quota.openshift.io/v1.ClusterResourceQuota.html) custom resource, and code in the `kubernetes_apiserver` to report metrics on it.

For each resource type, the agent will report total and per-namespace gauges. For example with CPU on the [test fixture](https://github.com/DataDog/datadog-agent/blob/22b5bdb26de9f76553fe5917b5acc63253473845/pkg/collector/corechecks/cluster/testdata/oshift_crq_list.json), we will report:

With the tag `clusterquota:multiproj-test`:
- `openshift.clusterquota.cpu.used` = 0.6 (cores)
- `openshift.clusterquota.cpu.limit` = 3 (cores)
- `openshift.clusterquota.cpu.remaining` = 2.4 (cores): computed as `limit - used`

These are to be used per quota object, and for cluster-wide sums

With the tags `clusterquota:multiproj-test kube_namespace:proj2`
- `openshift.appliedclusterquota.cpu.used` = 0 (cores)
- `openshift.appliedclusterquota.cpu.limit` = 3 (cores)
- `openshift.appliedclusterquota.cpu.remaining` = 2.4 (cores): we're using 0, but we already use 0.6 in another namespace

The are only to be used at the namespace level (in a scoped dashboard), the cluster-wide aggregates will be off. If several quotas apply to a given namespace, the min aggregation of `openshift.appliedclusterquota.cpu.remaining` will indicate what can be used before hitting the least permissive one.

### Additional Notes

- Test images `datadog/agent-dev:xvello-oshift-quotas` (and `-jmx`) are pushed on docker hub
- Legacy `OAPI` compatibility mode still needs to be tested.

xvello committed:
add openshift clusterquota metrics

xvello commented:
Pinging @burningion for inclusion in the OShift doc once we finalise the metric names 

@masci What's the best place to store metric metadata for these new metrics? Should we whitelist the `openshift.` prefix too?

xvello committed:
reno

xvello committed:
fix oapi endpoint, add oshift detection to diagnose

xvello commented:
Testing OK on OpenShift 3.4 too, metric metadata from https://github.com/DataDog/integrations-core/pull/1902 are applied fine:
![](https://cl.ly/0s3F1d3E2C3f/Image%202018-07-18%20at%206.48.47%20PM.png)

devonboyer commented on Dockerfiles/manifests/rbac/clusterrole.yaml:
```diff
@@ -16,6 +16,12 @@ rules:
   - get
   - list
   - watch
+- apiGroups: ["quota.openshift.io"]
```
Should this be commented out here? What happens if someone tries to create this manifest that isn't using openshift?

xvello replied:
Unknown `apigroups` or `resources` are silently ignored, so we're not triggering an error on vanilla k8s.

Indeed, we could split these in a separate `clusterrole`+`clusterrolebinding` pair on the same `serviceaccount`, I'm concerned about lengthening the setup instructions though.

devonboyer commented on pkg/util/kubernetes/apiserver/openshift.go:
```diff
@@ -0,0 +1,65 @@
+	if c.isOpenShift != OpenShiftUnknown {
+		return c.isOpenShift
+	}
+	err := c.Cl.CoreV1().RESTClient().Get().AbsPath("/apis/quota.openshift.io").Do().Error()
```
What about using the `Discovery` api here to see what OpenShift group versions are available?

https://github.com/kubernetes/client-go/blob/master/discovery/discovery_client.go#L135

CharlyF replied:
Agreed, has this path changed ? Will it potentially change ?

xvello replied:
Discovery won't play well with the legacy `/oapi/` endpoint as it's openshift-specific. I went with the same logic for both endpoints to have the same code. Both endpoints are documented in 3.10:

- https://docs.openshift.org/latest/rest_api/apis-quota.openshift.io/v1.ClusterResourceQuota.html
- https://docs.openshift.org/latest/rest_api/oapi/v1.ClusterResourceQuota.html

Redhat being Redhat, these endpoints are here to stay a *long* time. The "old" one is not even listed as deprecated, we could even just use this one unconditionally. I went with the new one when available to be future-proof and make the rbac more idiomatic though.

devonboyer commented on pkg/util/kubernetes/apiserver/openshift.go:
```diff
@@ -0,0 +1,65 @@
+	case OpenShiftOApi:
+		url = "/oapi/v1/clusterresourcequotas/"
+		break
+	}
```
What about if we keep a slice of all available OpenShift group versions instead? It's possible for more than one group version to be available at a time.
 
Then we could make the group version being used here configurable. The default could always be the latest group version `/apis/quota.openshift.io/v1/clusterresourcequotas/`.

devonboyer commented on pkg/util/kubernetes/apiserver/openshift.go:
```diff
@@ -0,0 +1,65 @@
+var ErrNotOpenShift = errors.New("not an OpenShift cluster")
+
+// IsOpenShift detects available endpoints to determine of OpenShift APIs are available
+func (c *APIClient) IsOpenShift() OpenShiftApiLevel {
```
Maybe instead this could be `GetOpenShiftGroupVersions() []string` and `IsOpenShift` could just return `true` if any OpenShift group version is available?

CharlyF commented on pkg/util/kubernetes/apiserver/openshift.go:
```diff
@@ -0,0 +1,65 @@
+// from the APIserver, returns ErrNotOpenShift if called on a non-Openshift cluster
+func (c *APIClient) ListOShiftClusterQuotas() ([]osq.ClusterResourceQuota, error) {
+	var url string
+	switch c.IsOpenShift() {
```
ListOShiftClusterQuotas is called after evaluating IsOpenShift(), is there any way we can factorize ?
Trying to reduce the number of calls to the API Server

xvello replied:
I do agree `c.isOpenShift` is not very elegant. WDYT about keeping this in the check itself, and even move `ListOShiftClusterQuotas` there, as I don't foresee other uses for it? `GetRESTObject` would stay under apiclient, as it can have other future uses.

CharlyF replied:
I am a fan of having a dedicated openshift.go file, I think all the relevant methods should be there. 👍 for the `GetRESTObject` under apiClient. If there are other things to list in the next versions of OpenShift, they could go alongside `ListOShiftClusterQuotas`.
Would that make sense ?

CharlyF reviewed (commented):
Left a few minor comments. I was about to ask if there is a way we can encapsulate the APIClient if we are using OpenShift to avoid adding a vendor specific boolean in the struct, but I could not fund good examples, and the "official" client has not been updated for a while https://github.com/openshift/kubernetes-client-go

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -54,6 +54,7 @@ type KubeASCheck struct {
 func (c *KubeASConfig) parse(data []byte) error {
 	// default values
 	c.CollectEvent = config.Datadog.GetBool("collect_kubernetes_events")
+	c.CollectOShiftQuotas = true
```
This is the default value (but we do need the `omitempty` for that to work, good catch).
If the api is not available, the collection is silently skipped, see https://github.com/DataDog/datadog-agent/pull/1945/files#diff-49e8a5417c408bfd4b45b146448dd0c0R124

Resource usage is pretty light (it's just one `list` call per check run), that's why I went with enabling it by default. WDYT?

xvello committed:
PR feedback

- Move ListOShiftClusterQuotas to check itself, as it's the only user
- Don't keep isOpenShift in the apiclient, move it to check state
- Improve comments

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -36,9 +35,10 @@ const (
-	FilteredEventType []string `yaml:"filtered_event_types"`
+	Tags                []string `yaml:"tags"`
+	CollectEvent        bool     `yaml:"collect_events"`
+	CollectOShiftQuotas bool     `yaml:"collect_openshift_clusterquotas"`
```
From my understanding [of the docs](https://godoc.org/gopkg.in/yaml.v2), `omitempty` is only used when marshalling. When unmarshalling, missing yaml fields won't trigger an error, and the struct fields will be left to their current value (that's how we set the defaults before unmarshalling)

CharlyF commented on pkg/collector/corechecks/cluster/kubernetes_openshift.go:
```diff
@@ -0,0 +1,100 @@
+	var url string
+	switch k.oshiftAPILevel {
+	case apiserver.OpenShiftAPIGroup:
+		url = "/apis/quota.openshift.io/v1/clusterresourcequotas/"
```
Small nit, you can ignore but we could use a `var baseUrl  = "/apis/quota.openshift.io"`  `var oapiUrl = "/oapi"` and then `var clusterQuotaEndpoint = "/v1/clusterresourcequotas/"` ?

xvello committed:
move endpoints to contants

CharlyF reviewed (approved):
🌵

devonboyer commented on pkg/util/kubernetes/apiserver/openshift_types.go:
```diff
@@ -0,0 +1,18 @@
+package apiserver
+
+// OpenShiftAPILevel describes what level of OpenShift APIs are available on the apiserver
+type OpenShiftAPILevel string
```
Does `OpenShift` have any other apigroups/custom resources other than clusterquotas we may want to collect metrics for in the future?

Should this maybe be called `OpenShiftQuotasAPILevel`

```
OpenShiftQuotasAPIGroup
OpenShiftQuotasOAPI
OpenShiftQuotasNone
```

devonboyer commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -484,3 +485,13 @@ func GetResourcesNamespace() string {
+
+// GetRESTObject allows to retrive a custom resource from the APIserver
+func (c *APIClient) GetRESTObject(path string, output runtime.Object) error {
+	result := c.Cl.CoreV1().RESTClient().Get().AbsPath(path).Do()
```
You can actually do this in one line like

```
c.Cl.CoreV1().RESTClient().Get().AbsPath(path).Do().Into(result).Error()
```

If an error occurred `Into` will simply be a noop and then `Error` will return any error or `nil`.
